### ISSUE TITLE
Variadic constructors

### ISSUE THREAD

**Issue by ds84182** [dart-lang/language#305]
[Slightly outdated proposal document (from May 2018!)](https://gist.github.com/ds84182/3d54ed9992aaba8e82be556a155fe277)

In short, variadic constructors would allow custom collection types (incl TypedData) to be created directly from a lazy evaluated iterable or list (depending on the type specified in the constructor).

This would be great for initializing typed data, and could also be extended to allow for some restricted constness (const typed data!!). It could also make Flutter build methods slightly more concise:

```dart
Widget build(BuildContext context) {
  return Row [
    Text("foo"),
    Text("bar"),
  ];
}
```

The only downside is that named parameters can't be passed to Row using this syntax.

Perhaps a different approach can be taken for defining the variadic constructors themselves (like `class Foo<E> { Foo[E data...] {} }`.

Maps could also benefit from this syntax, so custom map collection types can be initialized using the familiar map syntax: `BuiltMap { "foo": bar }`

**Comment by munificent** [comment]
That syntax wouldn't work because it collides with index operators. Consider a row containing only a single child widget:

```dart
Row[Text("foo")];
```

So I don't think we'd want that exact syntax. However, we could do something like [rest parameters](https://github.com/munificent/ui-as-code/blob/master/in-progress/parameter-freedom.md#rest-parameters).

**Comment by ds84182** [comment]
Ah, that's true. However, you can't use the index operator on a type, so there shouldn't be any existing code that conflicts, right? Optional new works today, so it would seem like the syntax could work..

**Comment by munificent** [comment]
> However, you can't use the index operator on a type, so there shouldn't be any existing code that conflicts, right?

Yes, but we want the grammar to be context-free. We don't want to have to resolve types in order to know how to *parse* a file. Needing symbol resolution during parsing is a road to madness and heartbreak.

(For example, consider that this means you would be unable to run dartfmt on a file containing a type error.)

**Comment by mraleph** [comment]
> We don't want to have to resolve types in order to know how to parse a file.

@munificent note that `Row()` parses as a constructor invocation if `Row` is a type and as a method invocation otherwise. This means that parser already has to resolve `Row` to select that parse at some point.

**Comment by munificent** [comment]
> parses as a constructor invocation if Row is a type and as a method invocation otherwise.

Sure, but those two are syntactically identical so the parser doesn't really care. Especially once you get into static methods, named constructors, and factory constructors, I think there's an argument that all invocations are syntactically just "invocation". Consider:

```dart
import "mystery.dart";

var x = fromMystery(1, 2);
var y = fromMystery.another(1, 2);
```

A Dart parser is able to parse those without having to load `mystery.dart`, even though neither, either, or both of those could be constructors.

This is in contrast with:

```dart
import "mystery.dart";

main() {
  fromMystery[1] = 2;
}
```

If `fromMystery` is a list object, that should be fine. If it's a type, I would expect that to be a syntax error because the LHS of an assignment can't be a constructor invocation.

**Comment by mraleph** [comment]
> If it's a type, I would expect that to be a syntax error because the LHS of an assignment can't be a constructor invocation.

I guess we are just slightly disagreeing on what _to parse_ means and what is the result of this parsing.

Result of your parsing is _concrete syntax tree_ of an input Dart program, which still needs to go through _resolution_ before it becomes something informative enough for backends and other consumers. 

When I am talking about parsing I assume that we produce a (resolved) abstract syntax tree - which seems what most consumers want anyway. 

Do any consumers actually care about _concrete syntax trees_? If we do have such consumers then it indeed makes sense to design grammar in such a way that its concrete syntax trees are somewhat meaningful.

**Comment by munificent** [comment]
> Do any consumers actually care about concrete syntax trees? 

Yes, dartfmt does not resolve identifiers and we want to support formatting code that contains resolution errors. In particular, if you want to be able to parse a single file without looking at its imports, that implies that you aren't resolving identifiers either since many of those may come from other libraries.

**Comment by lrhn** [comment]
If the formatting of a concrete syntax tree is always the same, independently of how identifiers resolve, then you *can* do the formatting without knowing whether `Row[1]` is an index operator invocation or "row literal". It's just `identifier` followed by `index operator` in the concrete syntax.

That would put some restrictions on how we can format things, though.

It does mean that there is a larger distance between the concrete syntax tree and the resolved abstract syntax tree. (We already have to handle that with `p.o.x(4)` potentially being a single constructor invocation and potentially being one getter invocation and one method invocation, so it's not unprecedented. We do format those two exactly the same, even though we could, say, have had rules saying that a prefix is never split from its member).

**Comment by rrousselGit** [comment]
> ```dart
> Row of [
>   Column of [
>     if (x) Foo(),
>     Bar(),
>   ],
> ];
> ```

What's the benefits vs:

```
Row.of([
  Column.of([
    if (x) Foo(), 
    Bar(), 
  ])
])
```
?
